File: packages/diagnostic-client/src/evidence.ts
import {
  ArtifactEvidenceSchema,
  LogQueryEvidenceSchema,
  MetricQueryEvidenceSchema,
  ProfileEvidenceSchema,
  SourceEvidenceSchema,
  TraceEvidenceSchema,
  type DiagnosticEvidence,
  type TimeRange,
} from "@moesegfault/contracts";
import { sanitizeText } from "@moesegfault/telemetry";

/** Locator query 可携带的安全标量。/ Safe scalar accepted in locator queries. */
export type SafeQueryValue = string | number | boolean;

/** 日志 locator 的固定低基数键。/ Fixed low-cardinality keys for log locators. */
export type LogQueryKey =
  | "service"
  | "environment"
  | "deployment_id"
  | "trace_id"
  | "span_id"
  | "severity"
  | "operation"
  | "component"
  | "error_type";

/** Profile locator 的固定低基数键。/ Fixed low-cardinality keys for profile locators. */
export type ProfileQueryKey =
  "service" | "environment" | "deployment_id" | "region" | "instance_id";

/** Metric locator 的固定低基数键。/ Fixed low-cardinality keys for metric locators. */
export type MetricQueryKey =
  | "service"
  | "environment"
  | "deployment_id"
  | "region"
  | "operation"
  | "component"
  | "dependency";

declare const SAFE_EVIDENCE: unique symbol;

/** 由有限 builder 产生的安全证据联合。/ Safe evidence union produced by finite builders. */
export type SafeDiagnosticEvidence = DiagnosticEvidence & {
  readonly [SAFE_EVIDENCE]: true;
};

const LOG_KEYS = new Set<LogQueryKey>([
  "service",
  "environment",
  "deployment_id",
  "trace_id",
  "span_id",
  "severity",
  "operation",
  "component",
  "error_type",
]);
const PROFILE_KEYS = new Set<ProfileQueryKey>([
  "service",
  "environment",
  "deployment_id",
  "region",
  "instance_id",
]);
const METRIC_KEYS = new Set<MetricQueryKey>([
  "service",
  "environment",
  "deployment_id",
  "region",
  "operation",
  "component",
  "dependency",
]);

/** 构建 trace locator，不接受 URL、headers 或自由查询。/ Builds a trace locator without URLs, headers, or free-form queries. */
export function traceEvidence(
  input: Readonly<{
    backend: string;
    traceId: string;
    spanId?: string;
  }>,
): SafeDiagnosticEvidence {
  return safe(
    TraceEvidenceSchema.parse({
      kind: "trace",
      backend: input.backend,
      locator: {
        trace_id: input.traceId,
        ...(input.spanId === undefined ? {} : { span_id: input.spanId }),
      },
    }),
  );
}

/** 构建固定键日志查询 locator。/ Builds a fixed-key structured-log query locator. */
export function logQueryEvidence(
  input: Readonly<{
    backend: string;
    query: Readonly<Partial<Record<LogQueryKey, SafeQueryValue>>>;
    timeRange: TimeRange;
  }>,
): SafeDiagnosticEvidence {
  return safe(
    LogQueryEvidenceSchema.parse({
      kind: "log_query",
      backend: input.backend,
      locator: { query: safeQuery(input.query, LOG_KEYS) },
      time_range: input.timeRange,
    }),
  );
}

/** 构建固定键 profile locator。/ Builds a fixed-key continuous-profile locator. */
export function profileEvidence(
  input: Readonly<{
    backend: string;
    profileType:
      "cpu" | "memory" | "allocations" | "mutex" | "goroutine" | "wall";
    profileId?: string;
    query?: Readonly<Partial<Record<ProfileQueryKey, SafeQueryValue>>>;
    timeRange: TimeRange;
  }>,
): SafeDiagnosticEvidence {
  return safe(
    ProfileEvidenceSchema.parse({
      kind: "profile",
      backend: input.backend,
      locator: {
        profile_type: input.profileType,
        ...(input.profileId === undefined
          ? {}
          : { profile_id: sanitizeRequired(input.profileId, 256) }),
        ...(input.query === undefined
          ? {}
          : { query: safeQuery(input.query, PROFILE_KEYS) }),
      },
      time_range: input.timeRange,
    }),
  );
}

/** 构建固定键 metric locator。/ Builds a fixed-key metric-query locator. */
export function metricQueryEvidence(
  input: Readonly<{
    backend: string;
    metricName: string;
    query: Readonly<Partial<Record<MetricQueryKey, SafeQueryValue>>>;
    timeRange: TimeRange;
  }>,
): SafeDiagnosticEvidence {
  return safe(
    MetricQueryEvidenceSchema.parse({
      kind: "metric_query",
      backend: input.backend,
      locator: {
        metric_name: sanitizeRequired(input.metricName, 255),
        query: safeQuery(input.query, METRIC_KEYS),
      },
      time_range: input.timeRange,
    }),
  );
}

/** 构建固定 commit 的源码 locator，并拒绝 URL 凭据、query 与 fragment。/ Builds a commit-pinned source locator and rejects URL credentials, queries, and fragments. */
export function sourceEvidence(
  input: Readonly<{
    backend: string;
    repositoryUrl: string | URL;
    gitCommit: string;
    path: string;
    line?: number;
    column?: number;
  }>,
): SafeDiagnosticEvidence {
  const repositoryUrl = safeRepositoryUrl(input.repositoryUrl);
  return safe(
    SourceEvidenceSchema.parse({
      kind: "source",
      backend: input.backend,
      locator: {
        repository_url: repositoryUrl,
        git_commit: input.gitCommit,
        path: sanitizeRequired(input.path, 1_024),
        ...(input.line === undefined ? {} : { line: input.line }),
        ...(input.column === undefined ? {} : { column: input.column }),
      },
    }),
  );
}

/** 构建内容寻址的不可变产物 locator。/ Builds a content-addressed immutable-artifact locator. */
export function artifactEvidence(
  input: Readonly<{
    backend: string;
    artifactDigest: `sha256:${string}`;
    artifactKind:
      "binary" | "debug_symbols" | "source_map" | "sbom" | "manifest" | "other";
    buildId?: string;
  }>,
): SafeDiagnosticEvidence {
  return safe(
    ArtifactEvidenceSchema.parse({
      kind: "artifact",
      backend: input.backend,
      locator: {
        artifact_digest: input.artifactDigest,
        artifact_kind: input.artifactKind,
        ...(input.buildId === undefined
          ? {}
          : { build_id: sanitizeRequired(input.buildId, 256) }),
      },
    }),
  );
}

/**
 * 在最终事件边界重建证据，阻止类型断言绕过有限 locator API。
 * Rebuilds evidence at the final event boundary so type assertions cannot bypass the finite locator API.
 *
 * @internal
 */
export function validateSafeEvidence(
  input: SafeDiagnosticEvidence,
): SafeDiagnosticEvidence {
  switch (input.kind) {
    case "trace":
      return traceEvidence({
        backend: input.backend,
        traceId: input.locator.trace_id,
        ...(input.locator.span_id === undefined
          ? {}
          : { spanId: input.locator.span_id }),
      });
    case "log_query":
      return logQueryEvidence({
        backend: input.backend,
        query: input.locator.query,
        timeRange: input.time_range,
      });
    case "profile":
      return profileEvidence({
        backend: input.backend,
        profileType: input.locator.profile_type,
        ...(input.locator.profile_id === undefined
          ? {}
          : { profileId: input.locator.profile_id }),
        ...(input.locator.query === undefined
          ? {}
          : { query: input.locator.query }),
        timeRange: input.time_range,
      });
    case "metric_query":
      return metricQueryEvidence({
        backend: input.backend,
        metricName: input.locator.metric_name,
        query: input.locator.query,
        timeRange: input.time_range,
      });
    case "source":
      return sourceEvidence({
        backend: input.backend,
        repositoryUrl: input.locator.repository_url,
        gitCommit: input.locator.git_commit,
        path: input.locator.path,
        ...(input.locator.line === undefined
          ? {}
          : { line: input.locator.line }),
        ...(input.locator.column === undefined
          ? {}
          : { column: input.locator.column }),
      });
    case "artifact":
      return artifactEvidence({
        backend: input.backend,
        artifactDigest: input.locator.artifact_digest as `sha256:${string}`,
        artifactKind: input.locator.artifact_kind,
        ...(input.locator.build_id === undefined
          ? {}
          : { buildId: input.locator.build_id }),
      });
  }
}

/** 仅复制显式固定键并清理字符串值。/ Copies only explicit fixed keys and scrubs string values. */
function safeQuery<K extends string>(
  input: Readonly<Partial<Record<K, SafeQueryValue>>>,
  allowed: ReadonlySet<K>,
): Readonly<Record<string, SafeQueryValue>> {
  const output: Record<string, SafeQueryValue> = Object.create(null) as Record<
    string,
    SafeQueryValue
  >;
  for (const key of Object.keys(input).sort()) {
    if (!allowed.has(key as K)) {
      throw new TypeError(`locator query key is not allowed: ${key}`);
    }
    const value = input[key as K];
    if (typeof value === "string") output[key] = sanitizeRequired(value, 512);
    else if (typeof value === "boolean") output[key] = value;
    else if (typeof value === "number" && Number.isFinite(value))
      output[key] = value;
    else throw new TypeError(`locator query value is invalid: ${key}`);
  }
  if (Object.keys(output).length === 0) {
    throw new TypeError("locator query requires at least one fixed key");
  }
  return Object.freeze(output);
}

/** 清理必须保持非空的 locator 文本。/ Scrubs locator text that must remain non-empty. */
function sanitizeRequired(value: string, maxLength: number): string {
  const safe = sanitizeText(value, maxLength);
  if (safe.length === 0) throw new TypeError("locator text must not be empty");
  return safe;
}

/** 解析安全的源码仓库 URL。/ Parses a safe source-repository URL. */
function safeRepositoryUrl(value: string | URL): string {
  let url: URL;
  try {
    url = new URL(value.toString());
  } catch {
    throw new TypeError("repositoryUrl must be an absolute HTTPS URL");
  }
  if (
    url.protocol !== "https:" ||
    url.username !== "" ||
    url.password !== "" ||
    url.search !== "" ||
    url.hash !== ""
  ) {
    throw new TypeError(
      "repositoryUrl must use HTTPS and contain no credentials, query, or fragment",
    );
  }
  if (sanitizeText(url.href, 2_048) !== url.href) {
    throw new TypeError("repositoryUrl contains a credential-like value");
  }
  return url.href;
}

/** 深冻结由 Zod 克隆后的 JSON 值。/ Deep-freezes the JSON value cloned by Zod. */
function freeze<T>(value: T): T {
  if (typeof value !== "object" || value === null || Object.isFrozen(value))
    return value;
  for (const child of Object.values(value)) freeze(child);
  return Object.freeze(value);
}

/** 标记并冻结已重建的安全证据。/ Brands and freezes rebuilt safe evidence. */
function safe(value: DiagnosticEvidence): SafeDiagnosticEvidence {
  return freeze(value) as SafeDiagnosticEvidence;
}

File: crates/diagnostic-client/src/evidence.rs
//! 有限、不可伪造的证据构建器。 / Finite unforgeable evidence builders.
use crate::propagation::valid_hex;
use serde_json::{json, Value};
use status_domain::{
    ArtifactKind, DiagnosticEvidence, DomainError, DomainResult, TelemetryKind, TimeRange,
};
use std::collections::BTreeMap;

/// 仅安全构建器可以构造。 / Constructible only through safe builders.
#[derive(Clone, Debug)]
pub struct SafeDiagnosticEvidence(DiagnosticEvidence);
impl SafeDiagnosticEvidence {
    /// 只读领域引用。 / Read-only domain reference.
    pub fn evidence(&self) -> &DiagnosticEvidence {
        &self.0
    }
}
/// 有限查询标量。 / Finite query scalar.
#[derive(Clone, Debug)]
pub enum SafeQueryValue {
    /// 分类文本而非用户输入。 / Classified text, never user input.
    Text(String),
    /// 有限数值。 / Finite number.
    Number(f64),
    /// 布尔分类。 / Boolean classification.
    Bool(bool),
}
/// 日志查询的固定键。 / Fixed log query keys.
#[derive(Clone, Copy, Debug, Ord, PartialOrd, Eq, PartialEq, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum LogQueryKey {
    /// 服务。 / Service.
    Service,
    /// 环境。 / Environment.
    Environment,
    /// 部署。 / Deployment.
    DeploymentId,
    /// Trace。 / Trace.
    TraceId,
    /// Span。 / Span.
    SpanId,
    /// 严重度。 / Severity.
    Severity,
    /// 操作。 / Operation.
    Operation,
    /// 组件。 / Component.
    Component,
    /// 错误类型。 / Error type.
    ErrorType,
}
/// Profile 查询固定键。 / Fixed profile query keys.
#[derive(Clone, Copy, Debug, Ord, PartialOrd, Eq, PartialEq, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ProfileQueryKey {
    /// 服务。 / Service.
    Service,
    /// 环境。 / Environment.
    Environment,
    /// 部署。 / Deployment.
    DeploymentId,
    /// 区域。 / Region.
    Region,
    /// 实例。 / Instance.
    InstanceId,
}
/// Metric 查询固定键。 / Fixed metric query keys.
#[derive(Clone, Copy, Debug, Ord, PartialOrd, Eq, PartialEq, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum MetricQueryKey {
    /// 服务。 / Service.
    Service,
    /// 环境。 / Environment.
    Environment,
    /// 部署。 / Deployment.
    DeploymentId,
    /// 区域。 / Region.
    Region,
    /// 操作。 / Operation.
    Operation,
    /// 组件。 / Component.
    Component,
    /// 依赖。 / Dependency.
    Dependency,
}
/// 有限 Profile 类别。 / Finite profile types.
#[derive(Clone, Copy, Debug, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ProfileType {
    /// CPU。 / CPU.
    Cpu,
    /// 内存。 / Memory.
    Memory,
    /// 分配。 / Allocations.
    Allocations,
    /// 锁。 / Mutex.
    Mutex,
    /// Goroutine。 / Goroutine.
    Goroutine,
    /// 墙钟时间。 / Wall time.
    Wall,
}
fn key_name<T: serde::Serialize>(key: T) -> String {
    serde_json::to_value(key)
        .expect("enum serialization")
        .as_str()
        .expect("string enum")
        .to_owned()
}
fn query<K: serde::Serialize + Ord>(input: BTreeMap<K, SafeQueryValue>) -> DomainResult<Value> {
    if input.is_empty() {
        return Err(invalid("query requires a fixed key"));
    }
    let mut result = BTreeMap::new();
    for (key, value) in input {
        let value = match value {
            SafeQueryValue::Text(s) => json!(required(&s, 512)?),
            SafeQueryValue::Bool(b) => json!(b),
            SafeQueryValue::Number(n) if n.is_finite() => json!(n),
            _ => return Err(invalid("query number must be finite")),
        };
        result.insert(key_name(key), value);
    }
    Ok(json!(result))
}
fn make(
    kind: TelemetryKind,
    backend: &str,
    locator: Value,
    time_range: Option<TimeRange>,
) -> DomainResult<SafeDiagnosticEvidence> {
    if backend.len() > 64
        || !regex::Regex::new(r"^[a-z0-9]+(?:[._-][a-z0-9]+)*$")
            .expect("constant regex")
            .is_match(backend)
    {
        return Err(invalid("invalid backend registry name"));
    }
    let evidence = DiagnosticEvidence {
        kind,
        backend: backend.into(),
        locator: serde_json::from_value(locator)?,
        time_range,
    };
    evidence.validate()?;
    Ok(SafeDiagnosticEvidence(evidence))
}
/// 构建 Trace 定位器。 / Builds a trace locator.
pub fn trace_evidence(
    backend: &str,
    trace_id: &str,
    span_id: Option<&str>,
) -> DomainResult<SafeDiagnosticEvidence> {
    if !valid_hex(trace_id, 32)
        || trace_id.bytes().all(|b| b == b'0')
        || span_id.is_some_and(|s| !valid_hex(s, 16) || s.bytes().all(|b| b == b'0'))
    {
        return Err(invalid("invalid trace identity"));
    }
    let mut locator = json!({"trace_id":trace_id});
    if let Some(s) = span_id {
        locator["span_id"] = json!(s);
    }
    make(TelemetryKind::Trace, backend, locator, None)
}
/// 构建日志查询。 / Builds a log query.
pub fn log_query_evidence(
    backend: &str,
    values: BTreeMap<LogQueryKey, SafeQueryValue>,
    range: TimeRange,
) -> DomainResult<SafeDiagnosticEvidence> {
    make(
        TelemetryKind::LogQuery,
        backend,
        json!({"query":query(values)?}),
        Some(range),
    )
}
/// 构建 Profile 查询或对象引用。 / Builds a profile query or object reference.
pub fn profile_evidence(
    backend: &str,
    kind: ProfileType,
    id: Option<&str>,
    values: Option<BTreeMap<ProfileQueryKey, SafeQueryValue>>,
    range: TimeRange,
) -> DomainResult<SafeDiagnosticEvidence> {
    if id.is_none() && values.is_none() {
        return Err(invalid("profile requires ID or query"));
    }
    let mut locator = json!({"profile_type":key_name(kind)});
    if let Some(id) = id {
        locator["profile_id"] = json!(required(id, 256)?);
    }
    if let Some(q) = values {
        locator["query"] = query(q)?;
    }
    make(TelemetryKind::Profile, backend, locator, Some(range))
}
/// 构建指标查询。 / Builds a metric query.
pub fn metric_query_evidence(
    backend: &str,
    name: &str,
    values: BTreeMap<MetricQueryKey, SafeQueryValue>,
    range: TimeRange,
) -> DomainResult<SafeDiagnosticEvidence> {
    make(
        TelemetryKind::MetricQuery,
        backend,
        json!({"metric_name":required(name,255)?,"query":query(values)?}),
        Some(range),
    )
}
/// 构建固定 commit 的无凭据源码引用。 / Builds a credential-free commit-pinned source reference.
pub fn source_evidence(
    backend: &str,
    repository: &str,
    commit: &str,
    path: &str,
    line: Option<u32>,
    column: Option<u32>,
) -> DomainResult<SafeDiagnosticEvidence> {
    let url = url::Url::parse(repository).map_err(|_| invalid("invalid repository URL"))?;
    if url.scheme() != "https"
        || !url.username().is_empty()
        || url.password().is_some()
        || url.query().is_some()
        || url.fragment().is_some()
        || sanitize_text(url.as_str(), 2048) != url.as_str()
    {
        return Err(invalid("repository must be safe HTTPS"));
    }
    if !(valid_hex(commit, 40) || valid_hex(commit, 64)) {
        return Err(invalid("commit must be full Git OID"));
    }
    let path = required(path, 1024)?;
    if path.starts_with('/')
        || path.split('/').any(|p| p == "..")
        || line
            .into_iter()
            .chain(column)
            .any(|n| n == 0 || n > 10_000_000)
    {
        return Err(invalid("invalid source position"));
    }
    let mut locator = json!({"repository_url":url.as_str(),"git_commit":commit,"path":path});
    if let Some(n) = line {
        locator["line"] = json!(n);
    }
    if let Some(n) = column {
        locator["column"] = json!(n);
    }
    make(TelemetryKind::Source, backend, locator, None)
}
/// 构建内容寻址产物引用。 / Builds a content-addressed artifact reference.
pub fn artifact_evidence(
    backend: &str,
    digest: &str,
    kind: ArtifactKind,
    build_id: Option<&str>,
) -> DomainResult<SafeDiagnosticEvidence> {
    if !digest
        .strip_prefix("sha256:")
        .is_some_and(|v| valid_hex(v, 64))
    {
        return Err(invalid("invalid artifact digest"));
    }
    let mut locator = json!({"artifact_digest":digest,"artifact_kind":kind});
    if let Some(id) = build_id {
        locator["build_id"] = json!(required(id, 256)?);
    }
    make(TelemetryKind::Artifact, backend, locator, None)
}
/// 清理常见凭据形态；不是通用秘密检测器。 / Scrubs common credential shapes, not arbitrary secrets.
pub(crate) fn sanitize_text(value: &str, limit: usize) -> String {
    use std::sync::OnceLock;
    static SECRET: OnceLock<regex::Regex> = OnceLock::new();
    static JWT: OnceLock<regex::Regex> = OnceLock::new();
    let secret=SECRET.get_or_init(||regex::Regex::new(r"(?i)\b(?:bearer|basic)\s+[a-z0-9._~+\-/]+=*|\b(?:password|passwd|secret|token|api[_-]?key|authorization|cookie)\s*[:=]\s*[^\s,;]+|\b[a-z][a-z0-9+.-]*://[^\s/@:]+:[^\s/@]+@").expect("constant regex"));
    let jwt = JWT.get_or_init(|| {
        regex::Regex::new(r"\beyJ[a-zA-Z0-9_-]{8,}\.[a-zA-Z0-9_-]{8,}\.[a-zA-Z0-9_-]{8,}\b")
            .expect("constant regex")
    });
    let first = secret.replace_all(value, "[REDACTED]");
    let clean = jwt.replace_all(&first, "[REDACTED]");
    if clean.encode_utf16().count() <= limit {
        return clean.into_owned();
    }
    let mut result = String::new();
    let mut units = 0;
    for c in clean.chars() {
        if units + c.len_utf16() >= limit {
            break;
        }
        units += c.len_utf16();
        result.push(c);
    }
    result.push('…');
    result
}
pub(crate) fn required(value: &str, limit: usize) -> DomainResult<String> {
    let s = sanitize_text(value, limit);
    if s.trim().is_empty() {
        Err(invalid("required text is empty"))
    } else {
        Ok(s)
    }
}
pub(crate) fn invalid(message: &str) -> DomainError {
    DomainError::Validation(message.into())
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn safe_locators_and_redaction() {
        assert!(source_evidence(
            "source",
            "https://a.example/?token=x",
            &"a".repeat(40),
            "a.rs",
            None,
            None
        )
        .is_err());
        assert!(artifact_evidence("r2", "sha256:bad", ArtifactKind::Other, None).is_err());
        assert_eq!(
            sanitize_text("Bearer abc token=xyz", 512),
            "[REDACTED] [REDACTED]"
        );
        assert!(trace_evidence("tempo", &"a".repeat(32), Some(&"0".repeat(16))).is_err());
    }
}

#[cfg(test)]
mod all_kinds_tests {
    use super::*;
    #[test]
    fn six_kinds_and_query_constraints() {
        let range = TimeRange {
            start: chrono::DateTime::from_timestamp_millis(0).unwrap(),
            end: chrono::DateTime::from_timestamp_millis(1).unwrap(),
        };
        let evidence = [
            trace_evidence("tempo", &"a".repeat(32), None).unwrap(),
            log_query_evidence(
                "logs",
                BTreeMap::from([(LogQueryKey::Service, SafeQueryValue::Text("api".into()))]),
                range.clone(),
            )
            .unwrap(),
            profile_evidence(
                "profiles",
                ProfileType::Cpu,
                Some("profile-1"),
                None,
                range.clone(),
            )
            .unwrap(),
            metric_query_evidence(
                "metrics",
                "latency",
                BTreeMap::from([(MetricQueryKey::Service, SafeQueryValue::Text("api".into()))]),
                range.clone(),
            )
            .unwrap(),
            source_evidence(
                "git",
                "https://example.com/repo",
                &"a".repeat(40),
                "src/lib.rs",
                Some(1),
                None,
            )
            .unwrap(),
            artifact_evidence(
                "r2",
                &format!("sha256:{}", "a".repeat(64)),
                ArtifactKind::Other,
                None,
            )
            .unwrap(),
        ];
        for e in evidence {
            e.evidence().validate().unwrap();
        }
        assert!(log_query_evidence("logs", BTreeMap::new(), range.clone()).is_err());
        assert!(profile_evidence("profiles", ProfileType::Cpu, None, None, range.clone()).is_err());
        assert!(metric_query_evidence(
            "metrics",
            "n",
            BTreeMap::from([(MetricQueryKey::Service, SafeQueryValue::Number(f64::NAN))]),
            range
        )
        .is_err());
    }
}
